pkg/endpoint/policy.go

```diff
@@ -435,7 +397,16 @@ func (e *Endpoint) updateRealizedState(stats *regenerationStatistics, origDir st
 	}
 
 	e.realizedBPFConfig = e.desiredBPFConfig
-	e.RealizedL4Policy = e.DesiredL4Policy
+
+	if e.realizedPolicy == nil {
+		e.realizedPolicy = &policy.EndpointPolicy{}
+	}
+
+	e.realizedPolicy.IngressPolicyEnabled = e.desiredPolicy.IngressPolicyEnabled
+	e.realizedPolicy.EgressPolicyEnabled = e.desiredPolicy.EgressPolicyEnabled
+	e.realizedPolicy.L4Policy = e.desiredPolicy.L4Policy
+	e.realizedPolicy.CIDRPolicy = e.desiredPolicy.CIDRPolicy
```

Comment:
Why is it necessary to reference individual pieces of the realized policy, rather than for example doing something like the following?
```
e.realizedPolicy = e.desiredPolicy
```
or 
```
e.realizedPolicy = e.desiredPolicy.Copy()
```

Reply:
The reason I did this is quite subtle (although it mirrors existing behavior), in that I do not set the references to be the same ( `e.realizedPolicy = e.desiredPolicy` ) because that would update the `MapState` fields in both objects to point to the same map. So, in `syncPolicyMap`, updating of the `desiredPolicy` `MapState` would also update the `realizedPolicy` `mapState`. The other fields in `desiredPolicy` are never updated except for during policy generation. 

We could `DeepCopy` all fields to avoid any potential corruption between fields, but I do wonder how that would impact performance, especially in the case of `L4Policy` with a large number of rules, which causes a lot of performance degradation due to the large number of `EndpointSelector` objects in `L4Filter.Endpoints`. 

What do you think?

Reply:
I don't see the point in making a deepcopy if we could ensure that, for example, the `L4Policy` object is not subsequently modified in the type of way you describe above.

I think all I'm really looking for here is to hide this `EndpointPolicy` field knowledge from the endpoint package. We could probably achieve this just by putting these lines behind a call that looks something like `e.realizedPolicy.Realizes(e.desiredPolicy)`:
```
// Realizes provides a way for the caller to update the received policy
// to reflect that it now has the same contents as the provided desired policy.
func (p *EndpointPolicy) Realizes(desired *EndpointPolicy) {
...
}
```

Reply:
@joestringer good idea - I'll add this to the follow-up ticket!